### PR TITLE
Enable M1 builds

### PR DESCRIPTION
**Pull request by Bouncner**
This PR makes some small changes to build Hyrise on Apple ARM.

Changes:
 * Clang (our only supported compiler on macOS) does not support `-march=native`
 * Some linker issues (`direct access in function [..] to global weak symbol 'typeinfo for [...] means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.`). @mrks had an idea how to fix it. Open for comments, if somebody has an idea how to address that in any different way.
 * Add a hint using the new homebrew path for ARM on M1 (`/opt/homebrew`) to the readline CMake file. Otherwise `libevent` (shipped with macOS) is used which does not compile due to missing functions.

### TPC-H SF1 comparison - left nemea, right M1
Please  note that nemea's CPU is three years old.
**NOT A PERFORMANCE COMPARISON FOR THE CHANGES OF THIS PR, BUT OF X86 and M1**

<details>
<summary>Configuration Overview - click to expand</summary>

```diff
 +Configuration Overview----+------------------------------------------------+------------------------------------------------+
 | Parameter                | ../rel_clang/master_1a73d9b4cd.json            | /home/Martin.Boissier/m1_master.json           |
 +--------------------------+------------------------------------------------+------------------------------------------------+
 |  GIT-HASH                | 1a73d9b4cd2d7b0cd49f8edff13af48f2357dbbf-dirty | e942ce92d166e448a5d6d45846d9b3a6dc693fdf-dirty |
 |  benchmark_mode          | Ordered                                        | Ordered                                        |
 |  build_type              | release                                        | release                                        |
 |  chunk_size              | 65535                                          | 65535                                          |
 |  clients                 | 1                                              | 1                                              |
 |  compiler                | clang 11.0.0                                   | clang 11.0.0                                   |
 |  cores                   | 0                                              | 0                                              |
 |  date                    | 2021-01-08 20:12:55                            | 2021-01-08 17:31:34                            |
 |  encoding                | {'default': {'encoding': 'Dictionary'}}        | {'default': {'encoding': 'Dictionary'}}        |
 |  indexes                 | False                                          | False                                          |
 |  max_duration            | 60000000000                                    | 60000000000                                    |
 |  max_runs                | -1                                             | -1                                             |
 |  scale_factor            | 1.0                                            | 1.0                                            |
 |  time_unit               | ns                                             | ns                                             |
 |  use_prepared_statements | False                                          | False                                          |
 |  using_scheduler         | False                                          | False                                          |
 |  verify                  | False                                          | False                                          |
 |  warmup_duration         | 0                                              | 0                                              |
 +--------------------------+------------------------------------------------+------------------------------------------------+
```
</details>

```diff
 +----------++----------+---------+--------++----------+----------+--------+---------+
 | Item     || Latency (ms/iter)  | Change || Throughput (iter/s) | Change | p-value |
 |          ||      old |     new |        ||      old |      new |        |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
+| TPC-H 01 ||    491.4 |   278.3 |  -43%  ||     2.04 |     3.59 |  +77%  |  0.0000 |
+| TPC-H 02 ||      4.7 |     4.3 |   -9%  ||   211.07 |   232.05 |  +10%  |  0.0000 |
+| TPC-H 03 ||     79.7 |    61.0 |  -23%  ||    12.54 |    16.39 |  +31%  |  0.0000 |
+| TPC-H 04 ||     71.3 |    58.7 |  -18%  ||    14.03 |    17.02 |  +21%  |  0.0000 |
+| TPC-H 05 ||    204.6 |   155.7 |  -24%  ||     4.89 |     6.42 |  +31%  |  0.0000 |
-| TPC-H 06 ||      3.2 |     5.6 |  +74%  ||   312.65 |   179.22 |  -43%  |  0.0000 |
+| TPC-H 07 ||     57.3 |    40.2 |  -30%  ||    17.46 |    24.88 |  +42%  |  0.0000 |
+| TPC-H 08 ||     58.5 |    47.7 |  -19%  ||    17.10 |    20.98 |  +23%  |  0.0000 |
+| TPC-H 09 ||    426.5 |   231.7 |  -46%  ||     2.34 |     4.32 |  +84%  |  0.0000 |
+| TPC-H 10 ||    108.4 |    77.3 |  -29%  ||     9.22 |    12.93 |  +40%  |  0.0000 |
+| TPC-H 11 ||      8.1 |     7.6 |   -6%  ||   123.16 |   131.14 |   +6%  |  0.0000 |
+| TPC-H 12 ||     44.0 |    33.1 |  -25%  ||    22.72 |    30.17 |  +33%  |  0.0000 |
+| TPC-H 13 ||    391.6 |   245.1 |  -37%  ||     2.55 |     4.08 |  +60%  |  0.0000 |
+| TPC-H 14 ||     20.7 |    14.8 |  -28%  ||    48.40 |    67.44 |  +39%  |  0.0000 |
+| TPC-H 15 ||      6.6 |     4.7 |  -29%  ||   152.37 |   214.47 |  +41%  |  0.0000 |
+| TPC-H 16 ||     70.1 |    49.8 |  -29%  ||    14.26 |    20.09 |  +41%  |  0.0000 |
-| TPC-H 17 ||     15.7 |    17.1 |   +9%  ||    63.56 |    58.53 |   -8%  |  0.0000 |
+| TPC-H 18 ||    464.3 |   267.3 |  -42%  ||     2.15 |     3.74 |  +74%  |  0.0000 |
 | TPC-H 19 ||     28.0 |    28.9 |   +3%  ||    35.73 |    34.63 |   -3%  |  0.0000 |
+| TPC-H 20 ||     15.0 |    12.5 |  -17%  ||    66.46 |    79.67 |  +20%  |  0.0000 |
+| TPC-H 21 ||    348.4 |   249.4 |  -28%  ||     2.87 |     4.01 |  +40%  |  0.0000 |
+| TPC-H 22 ||     30.7 |    20.1 |  -35%  ||    32.54 |    49.86 |  +53%  |  0.0000 |
 +----------++----------+---------+--------++----------+----------+--------+---------+
+| Sum      ||   2948.8 |  1911.0 |  -35%  ||          |          |        |         |
+| Geomean  ||          |         |        ||          |          |  +29%  |         |
 +----------++----------+---------+--------++----------+----------+--------+---------+
```

### CODE REVIEW

**Review by mrks — APPROVED** [review]

**Review by mrks — APPROVED** [review]

**Comment by mrks** [comment]
You might want to check if the unity builds make a difference.

**Comment by Bouncner** [comment]
A difference in what?

**Comment by mrks** [comment]
Never mind. It seems to work now. Did you figure out in which case we would need the new macro and in which case we can't use it? Haven't seen a pattern yet.

**Comment by Bouncner** [comment]
What are you looking at in particular? It's basically everything that also uses the explicitly-instantiate-macro. There are cases where we need to manually add them for some data types, because a class does not support all data types (e.g., FoR and strings), but that's the same for both macros.

Or do you mean something else?

**Review by mrks — COMMENTED** [review]
I just saw you adding and removing declarations and wondered what was going on there. Maybe this knowledge should be doc'ed somewhere?

**Comment by Bouncner** [comment]
A few of them have been just plain wrong. E.g., I added the declaration in the RLE encoder instead of the RLE segment.

File: src/lib/storage/index/b_tree/b_tree_index_impl.hpp [hunk 1]
```diff
@@ -64,4 +64,8 @@ class BTreeIndexImpl : public BaseBTreeIndexImpl {
   size_t _heap_bytes_used;
 };
 
+template<> void BTreeIndexImpl<pmr_string>::_add_to_heap_memory_usage(const pmr_string&);
```
[the remark this answers is not in the document]
Comment (Bouncner) [reply src/lib/storage/index/b_tree/b_tree_index_impl.hpp:67]: If we leave them out:
```../src/lib/storage/lz4_segment.cpp:128:49: error: explicit specialization of 'decompress' after instantiation
std::vector<pmr_string> LZ4Segment<pmr_string>::decompress() const {
                                                ^
/Users/martin/Programming/OpossumDB3/rel/../src/lib/storage/lz4_segment.hpp:225:31: note: explicit instantiation first required here
EXPLICITLY_DECLARE_DATA_TYPES(LZ4Segment);
                              ^
1 error generated.
```

If I put `extern` in front:
```
../src/lib/storage/lz4_segment.hpp:219:16: error: expected unqualified-id
extern template<> std::vector<pmr_string> LZ4Segment<pmr_string>::decompress() const;
```

**Review by mrks — APPROVED** [review]
Works for me. I would need more time to understand each and every added line, but as it doesn't change any "code", this works for me. Do you want to merge this?

**Pull request merged**